src/core/tsi/ssl_session.h

```diff
@@ -0,0 +1,73 @@
+/*
+ *
+ * Copyright 2018 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#ifndef GRPC_CORE_TSI_SSL_SESSION_H
+#define GRPC_CORE_TSI_SSL_SESSION_H
+
+#include <grpc/support/port_platform.h>
+
+#include <grpc/slice.h>
+
+extern "C" {
+#include <openssl/ssl.h>
```

Comment:
TSI header should not depend on ssl.h. TSI header also needs to be in C API. Therefore, please write in C. This is similar to the transport_security_interface.h.

E.g., 

```
typedef struct tsi_ssl_cached_session tsi_ssl_cached_session;

tsi_ssl_cached_session_create();
...
tsi_ssl_cached_session_destroy(tsi_ssl_cached_session self);
```

In ssl_session_boringssl.cc and ssl_session_openssl.cc implementations, you can still use c++. Later, you can write a wrapper to plumb the c++ implementation in the C interface.

Reply:
Cached session is not a part of TSI api, it's implementation detail of SslSessionCache.

Reply:
Should it & SslSessionCache C++ implementations be moved to `ssl/` subdirectory?